File: acm/Online-Judge/leetcode/code/3sum.py
class Solution:
    # @param {integer[]} nums
    # @return {integer[][]}
    def threeSum(self, num):
        mp={}
        n=len(num)
        ans=[]
        num.sort()
        l1=-3214421
        for i in xrange(n):
            if l1!=num[i]:
                l1=num[i]
                l2=-312414
                j=i+1
                p=n-1
                while j<p:
                    if l2!=num[j]:
                        l2=num[j]
                        while p>j and num[i]+num[j]+num[p]>0:
                            p=p-1
                        if p>j and num[i]+num[j]+num[p]==0:
                            ans.append([num[i],num[j],num[p]])
                    j=j+1
        return ans

File: acm/Online-Judge/leetcode/code/3sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int>& num)
    {
        vector<vector<int> > ans;
        sort(num.begin(),num.end());
        int n=num.size(),l1=-12314312;
        for(int i=0;i<n;i++)
            if(num[i]!=l1)
            {
                l1=num[i];
                int p=n-1,l2=-12314312;
                for(int j=i+1;j<p;j++)
                    if(num[j]!=l2)
                    {
                        l2=num[j];
                        while(p>j&&num[p]+num[j]+num[i]>0)
                            p--;
                        if(p>j&&num[p]+num[j]+num[i]==0)
                        {
                            vector<int>t;
                            t.push_back(num[i]);
                            t.push_back(num[j]);
                            t.push_back(num[p]);
                            ans.push_back(t);
                        }
                    }
            }
        return ans;
    }
};